jina/executors/decorators.py

```diff
@@ -365,3 +372,77 @@ def arg_wrapper(*args, **kwargs):
         return _batching(func)
     else:
         return _batching
+
+def batching_ranker_input(func: Callable[[Any], np.ndarray] = None,
+                         batch_size: Union[int, Callable] = None,
+                         num_batch: Optional[int] = None,
+                         slice_on: int = 2,
+                         num_data: int = 1) -> Any:
+    """Split the input of a function into small batches and call :func:`func` on each batch
+    , collect the merged result and return. This is useful when the input is too big to fit into memory.
+    This can batch multiple inputs. 
+
+    :param func: function to decorate
+    :param batch_size: size of each batch
+    :param num_batch: number of batches to take, the rest will be ignored
+    :param slice_on: the location of the data. When using inside a class,
+            ``slice_on`` should take ``self`` into consideration.
+    :param num_data: the number of data inside the arguments
+    :return: the merged result as if run :func:`func` once on the input.
+    ..warning:
+        data arguments will be taken starting from ``slice_on` to ``slice_on + num_data``
+    Example:
+        .. highlight:: python
+        .. code-block:: python
+            
+            class MemoryHungryRanker:
+                
+                @batching_ranker_input(batch_size = 64)
+                def score(
+                    self, query_meta: Dict, old_match_scores: Dict, match_meta: Dict
+                ) -> 'np.ndarray': 
+                ...       
+
+                    
+    """
+    def _batching(func):
+        @wraps(func)
+        def arg_wrapper(*args, **kwargs):
+            data = [args[slice_on + i] for i in range(0, num_data)]
+            # priority: decorator > class_attribute
+            # by default data is in args[1:] (self needs to be taken into account)
+            b_size = batch_size or getattr(args[0], 'batch_size', None)
+            # no batching if b_size is None
+            if b_size is None or data is None:
+                return func(*args, **kwargs)
+            args = list(args)
+            default_logger.debug(
+                f'batching enabled for {func.__qualname__} batch_size={b_size} '
+                f'num_batch={num_batch}')
+
+            # assume all datas have the same length
+            full_data_size = _get_size(args[slice_on])
+            total_size = _get_total_size(full_data_size, b_size, num_batch)
+            final_result = []
+            yield_dict = [isinstance(args[slice_on + i], Dict) for i in range(0,num_data)]
+            data_iterators = [batch_iterator(_get_slice(args[slice_on + i],total_size), b_size , yield_dict=yield_dict[i]) for i in range(0, num_data)]
```

Comment:
let's adapt this part together with `batching_multi_input` if possible. `get_slice` can be used also there right?. Also separating comma before `total_size`

Reply:
Since both `batch_multi_input` and `batch_ranker_input` are almost similar, we can completely remove `batch_ranker_input` and update `batching` and `batch_multi_input` with `get_slice` and `yield_dict` to handle dictionary data. or  keep 'batch_ranker_input' separate?

Reply:
yes lets try, but yield_dict can be avoided in regular batching. it can be set to default false in the batch_decorator

Reply:
if we are removing `batching_ranker_input` maybe we should add dictionary support to `batching` no?

Reply:
maybe in the future, but now I would prefer to keep it conservative and just touch where it may be used

Reply:
For rankers we actually need only one parameter `old_match_scores` to be batched. So we still need to use `batch_multi_input`  ?

Reply:
not true, you may want to batch also match meta